### ISSUE TITLE
Complete callback it's receiving partial results on long csv

### ISSUE THREAD

**Issue by topicus** [mholt/PapaParse#179]
I've tried to parse a long csv (1.8 millon of rows) and only the last chunk of parsed data it's received by the complete callback. 

### Steps to reproduce it.
- [ ] Go to http://papaparse.com/demo
- [ ] Download large csv file
- [ ] Use local file read method
- [ ] Press parse
- [ ] Row count is 179332 instead of 1094353 which is the real number of line rows.

**Comment by topicus** [comment]
I found the problematic commit using git bisect. 

https://github.com/mholt/PapaParse/commit/0785501b30b61aea9ab0a96dd2b8b8cdb9768260

I will try to fix it.

**Comment by mholt** [comment]
Oops! I guess we need tests for large files. I will try to find a good way to test those.

Thank you for looking into it on your own. :clap: I know I'm too busy until the semester ends to get around to it. Let me know if you have any questions along the way; I'll try to be of help.

**Comment by topicus** [comment]
I found the error but I'm not completely sure about the solution. https://github.com/mholt/PapaParse/blob/master/papaparse.js#L400

Just changing that line from:

```javascript
this._partialLine = aggregate.substring(lastIndex - this._baseIndex);
```
To this
```javascript
this._partialLine = aggregate.substring(this._baseIndex - lastIndex);
```
does the trick but I'm concerned about how this should work. In this way we are parsing over and over the same information. 

It would be better to have a sort of accumulator to store results.

I think FileStreamer was modified to parse by chunks but parseChunk wasn't prepared to allow that. 

I figured out this because parse method of ParseHandler constructor on the 4.0.7 version is executed only one time when I select a big file. By contrast, in the 4.1 version parse method is triggered as many times as chunks there are. 

Maybe you could make a brief explanation how this should work.

**Comment by mholt** [comment]
@bluej100 Would you have a chance to answer his questions? I would but school is keeping me way too busy right now.

**Comment by bluej100** [comment]
Yes, I will look into this tomorrow. Sorry.

**Comment by bluej100** [comment]
I've created a pull request. @mholt , I can leave it for your review (30 lines) or merge it as you prefer.

**Comment by bluej100** [comment]
By the way, @topicus , I appreciate that you dove into the code and suggested a fix. I went a different direction for now, restoring the previous behavior of scanning the entire file in one chunk, but I do like the accumulator as an idea for the future.

**Comment by mholt** [comment]
Thanks @bluej100 for picking up my slack :+1: I really appreciate it!

@topicus Does that PR (#181) fix the issue for you?

**Comment by topicus** [comment]
@bluej100 @mholt

Thanks for your quickly response, I've tried to parse the file but it's throwing "Uncaught RangeError: Invalid string length" when I use a big csv. 

I will take a look at the code during this weekend to help (if I can).

**Comment by bluej100** [comment]
Sorry to hear that. I'm surprised that it would have worked on the old code. I imagine that it's throwing the error on the reader.readAsText call?

It seems to me that for a file of this size you would probably be better off using a step or chunk function and streaming your results out. But adding an accumulator is a possibility.

What is the filesize of your test file?

**Comment by topicus** [comment]
Probably I tried with a slightly weightless file. File I'm currently trying to parse weight 259MB.
I've just created a PR with the accumulator approach. I don't completely sure if it will work for all the use cases. 

Maybe you could take a look and let me know if you see something wrong. 

https://github.com/mholt/PapaParse/pull/183

**Comment by bluej100** [comment]
We can close this now that #183 is merged, correct, @topicus ?

**Comment by JaroVDH** [comment]
When will this be put in a release?
Because right now we're also dealing with this issue.
Trying to use chunks to validate the parsed data is possible, but we do need the complete matrix at the end.
Is it worth it for me to just manually do some combining for now?

Also, do errors propagate through chunking? As in, do errors added in chunk's first callback call continue on to chunk's second callback call? Or does it just stop after the first error is added?

**Comment by mholt** [comment]
I'll do a release after I finish finals (next week) and have a chance to review all the issues. It's my fault that Papa Parse hasn't gotten much attention over the last 4 months - thanks to school. But the wait is nearly over.

You can manually combine chunks if you desire, but if the file is large the browser may choke on it.

Errors do not (last time I checked - and it should still be this way) persist through all the chunks. In other words, they are not cumulative. An error in chunk 0 will not appear in chunk 1 also. Each chunk is independent of another.

**Comment by JaroVDH** [comment]
Oh don't worry about Papa Parse too much if you've got finals to deal with. I'm sure everybody understands. They're much more important.

I ended up doing manual combining for now. Shouldn't be too much of a problem as files are rarely above 20MB. Same for errors, ended up combining them for each chunk to get a final tally of errors.

Thanks, and good luck with your finals.

**Comment by mholt** [comment]
@JaroVDH I released a patch 4.1.1 today which has the fix in it. It seems to be working for me, let me know if there's still a problem.
